nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-db-schema-registry-service/src/main/java/org/apache/nifi/db/schemaregistry/DatabaseSchemaRegistry.java

```diff
@@ -0,0 +1,179 @@
+        final Optional<String> schemaName = schemaIdentifier.getName();
+        if (schemaName.isEmpty()) {
+            throw new SchemaNotFoundException("Cannot retrieve schema because Schema Name is not present");
+        }
+
+        final String tableName = schemaName.get();
+        try {
+            try (final Connection conn = dbcpService.getConnection()) {
+                final DatabaseMetaData dmd = conn.getMetaData();
+                try (final ResultSet colrs = dmd.getColumns(dbCatalogName, dbSchemaName, tableName, "%")) {
+                    final List<RecordField> recordFields = new ArrayList<>();
+                    while (colrs.next()) {
+                        // COLUMN_DEF must be read first to work around Oracle bug, see NIFI-4279 for details
+                        final String defaultValue = colrs.getString("COLUMN_DEF");
+                        final String columnName = colrs.getString("COLUMN_NAME");
+                        final int dataType = colrs.getInt("DATA_TYPE");
+                        final String nullableValue = colrs.getString("IS_NULLABLE");
+                        final boolean isNullable = "YES".equalsIgnoreCase(nullableValue) || nullableValue.isEmpty();
+                        recordFields.add(new RecordField(
+                                columnName,
+                                DataTypeUtils.getDataTypeFromSQLTypeValue(dataType),
+                                defaultValue,
+                                isNullable));
+                    }
+
+                    // If no columns are found, check that the table exists
+                    if (recordFields.isEmpty()) {
+                        try (final ResultSet tblrs = dmd.getTables(dbCatalogName, dbSchemaName, tableName, null)) {
+                            List<String> qualifiedNameSegments = new ArrayList<>();
+                            if (dbCatalogName != null) {
+                                qualifiedNameSegments.add(dbCatalogName);
+                            }
+                            if (dbSchemaName != null) {
+                                qualifiedNameSegments.add(dbSchemaName);
+                            }
+                            qualifiedNameSegments.add(tableName);
+
+                            if (!tblrs.next()) {
+                                throw new SchemaNotFoundException("Table "
+                                        + String.join(".", qualifiedNameSegments)
+                                        + " not found");
+                            } else {
+                                getLogger().warn("Table "
+                                        + String.join(".", qualifiedNameSegments)
+                                        + " found but no columns were found, if this is not expected then check the user permissions for getting table metadata from the database");
+                            }
```

Comment:
To avoid deep levels of code nesting, what do you think about breaking this out to a separate method?